File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        Arrays.sort(time);
        long tt = totalTrips;
        long l = 1L;
        long h = (long)(time[time.length - 1] * tt);
        long m = (l + (h-l)/2); 
        long tripRec = 0L;
        long minTime = h;

        boolean verified = false;

        while(l <= h){
            m = l + (long)((h-l)/(long)2);

            verified = false;
            tripRec = 0L;

            for(int i=0;i<time.length;i++){
                if((long)time[i] > m || tripRec >= tt) break;
                tripRec += (long)m/(long)time[i];
            }
            if(tripRec >= tt){
                verified = true;
            }

            if(verified){
                minTime = (long)m;
                h = m - 1;
            }else{
                l = m + 1;
            }
        }
        return minTime;
    }
}

File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.cpp
class Solution {
public:
    long long minimumTime(vector<int>& time, int totalTrips) {
        sort(time.begin(), time.end());
        
        long long 
            l = 1, 
            h = (long long)time[time.size() - 1] * (long long)totalTrips, 
            m = l + (h-l)/2, 
            tripRec = 0, 
            minTime = h;
        
        bool verified = false;
        while(l <= h){
            m = l + (h-l)/2;
            
            verified = false;
            tripRec = 0;
            for(int i=0;i<time.size();i++){
                if(time[i] > m || tripRec >= totalTrips) break;
                tripRec += (long long)m / (long long)time[i];
            }
            if(tripRec >= totalTrips){
                verified = true;
            }
            
            if(verified){
                minTime = m;
                h = m - 1;
            }else{
                l = m + 1;
            }
        }
        return minTime;
    }
};